src/MonitorProfiler/MainProfiler/ExceptionTracker.cpp

```diff
@@ -0,0 +1,301 @@
+{
+    HRESULT hr = S_OK;
+
+    ObjectID thrownObjectId;
+    FunctionID catcherFunctionId;
+    IfFailLogRet(m_pThreadDataManager->GetException(threadId, &thrownObjectId, &catcherFunctionId));
+
+    if (ThreadData::NoFunctionId == catcherFunctionId)
+    {
+        tstring methodName;
+        IfFailLogRet(GetFullyQualifiedMethodName(functionId, methodName));
+        LogInformationV(_T("Exception unhandled: %s"), methodName.c_str());
+
+        // Future: Block thread until collection is initiated of the desired artifact.
+        // Possible serialization of some context of the exception and surrounding method
+        // information such as locals and parameters.
+    }
+    else if (functionId == catcherFunctionId)
+    {
+        IfFailLogRet(m_pThreadDataManager->ClearException(threadId));
+
+        // Exception handling is common; don't pay to calculate method name if it won't be logged.
+        if (m_pLogger->IsEnabled(LogLevel::Debug))
+        {
+            tstring methodName;
+            IfFailLogRet(GetFullyQualifiedMethodName(functionId, methodName));
+            LogDebugV(_T("Exception handled: %s"), methodName.c_str());
+        }
+    }
+
+    return S_OK;
+}
+
+HRESULT ExceptionTracker::MovedReferences(ULONG cMovedObjectIDRanges, ObjectID oldObjectIDRangeStart[], ObjectID newObjectIDRangeStart[], SIZE_T cObjectIDRangeLength[])
+{
+    HRESULT hr = S_OK;
+
+    IfFailLogRet(m_pThreadDataManager->MovedReferences(cMovedObjectIDRanges, oldObjectIDRangeStart, newObjectIDRangeStart, cObjectIDRangeLength));
+
+    return S_OK;
+}
+
+HRESULT ExceptionTracker::ExceptionThrownStackSnapshotCallback(
+    FunctionID funcId,
+    UINT_PTR ip,
+    COR_PRF_FRAME_INFO frameInfo,
+    ULONG32 contextSize,
+    BYTE context[],
+    void* clientData)
+{
+    // Callback is only used to get the function ID of the top frame.
+    *static_cast<FunctionID*>(clientData) = funcId;
+
+    // Cancel stack snapshot callbacks after the top frame.
+    return S_FALSE;
+}
+
+HRESULT ExceptionTracker::GetFullyQualifiedMethodName(FunctionID functionId, tstring& fullMethodName)
+{
+    HRESULT hr = S_OK;
+
+    ClassID classId;
+    ModuleID moduleId;
+    mdToken token;
+    IfFailLogRet(m_pCorProfilerInfo->GetFunctionInfo(
+        functionId,
+        &classId,
+        &moduleId,
+        &token
+    ));
+
+    if (0 == (token & mdtMethodDef))
+    {
+        return E_FAIL;
+    }
+    mdMethodDef methodDef = token;
+
+    ComPtr<IMetaDataImport2> pMetadataImport;
+    IfFailLogRet(m_pCorProfilerInfo->GetModuleMetaData(
+        moduleId,
+        CorOpenFlags::ofRead,
+        IID_IMetaDataImport2,
+        (IUnknown**)&pMetadataImport
+    ));
+
+    // Get Module Name
+    ULONG count = 0;
+    IfFailLogRet(m_pCorProfilerInfo->GetModuleInfo(
+        moduleId,
+        nullptr,
+        0,
+        &count,
+        nullptr,
+        nullptr
+    ));
+
+    unique_ptr<WCHAR[]> pModulePath(new (nothrow) WCHAR[count + 1]);
```

Comment:
I'm assuming the `+1` is to be null-terminated - and then I'm also assuming that `&count` will only be the number of _actual_ characters in the name, and is not null-terminated?

Reply:
The documentation is not exactly clear. I found other examples where the extra character is not necessary and some empirical evidence shows that it is not necessary. I've removed the extra character calculation and added comments that the null terminater is already accounted.